File: data-structure/data-structure/LinkedList.js
class Node {
    constructor(data) {
        this.data = data;
        this.next = null
    }
}

class LinkedList {
    constructor() {
        this.head = null;
        this.size = 0;
    }

    add(data) {
        let node = new Node(data);
        let current;

        if (this.head == null)
            this.head = node;
        else {
            current = this.head;

            while (current.next) {
                current = current.next;
            }
            current.next = node;
        }
        this.size++;
    }

    insertAt(data, index) {
        if (index > 0 && index > this.size)
            return false;
        else {
            let node = new Node(data);
            let curr;
            let prev;

            curr = this.head;

            if (index == 0) {
                node.next = this.head;
                this.head = node;
            } else {
                curr = this.head;
                let it = 0;

                while (it < index) {
                    it++;
                    prev = curr;
                    curr = curr.next;
                }
                node.next = curr;
                prev.next = node;
            }
            this.size++;
        }
    }

    remove(index) {
        if (index > 0 && index > this.size)
            return -1;
        else {
            let curr;
            let prev;
            let it = 0;
            curr = this.head;
            prev = curr;

            if (index === 0) {
                this.head = curr.next;
            } else {

                while (it < index) {
                    it++;
                    prev = curr;
                    curr = curr.next;
                }
                prev.next = curr.next;
            }
            this.size--;
            return curr.data;
        }
    }

    indexOf(data) {
        var count = 0;
        var current = this.head;

        while (current !== null) {
            if (current.data === data)
                return count;
            count++;
            current = current.next;
        }

        return -1;
    }

    isEmpty() {
        return this.size == 0;
    }

    length() {
        console.log(this.size);
    }

    printList() {
        var curr = this.head;
        var str = "";
        while (curr) {
            str += curr.data + " ";
            curr = curr.next;
        }
        console.log(str);
    }
}

const ll = new LinkedList();

console.log(ll.isEmpty()); // true
ll.add(10); // 10
ll.printList(); // 10
console.log(ll.length()); // 1
ll.add(20);
ll.add(30);
ll.add(40); 
ll.add(50); 
ll.printList(); // 10 20 30 40 50
console.log(ll.indexOf(40)); // 3
ll.insertAt(60, 2); 
ll.printList(); // 10 20 60 30 40 50
console.log(ll.isEmpty()); // false
console.log(ll.remove(3)); 
ll.printList(); // 10 20 60 40 50

File: data-structure/data-structure/LinkedList.py
# # Node 구현
# class Node:
#   def __init__(self, data, next=None):
#     self.data = data
#     self.next = next

# # 데이터 추가
# def add(data):
#   node = head
#   while node.next:
#     node = node.next
#   node.next = Node(data)

# # Node와 Node 연결 (포인터 활용)
# node1 = Node(1)
# head = node1
# for index in range(2, 10):
#   add(index)

# # 링크드 리스트 데이터 출력(검색)
# node = head
# while node.next:
#   print(node.data)
#   node = node.next
# print(node.data)


# # 데이터 사이에 데이터 추가
# node3 = Node(1.5)

# node = head
# search = True
# while search:
#   if node.data == 1:
#     search = False
#   else:
#     node = node.next

# node_next = node.next
# node.next = node3
# node3.next = node_next

# node = head
# while node.next:
#   print(node.data)
#   node = node.next
# print(node.data)

# 파이썬 객체지향으로 링크드 리스트 구현
class Node:
  def __init__(self, data, next=None):
    self.data = data
    self.next = next

class NodeMgmt:
  def __init__(self, data):
    self.head = Node(data)

  def add(self, data):
    if self.head == "":
      self.head = Node(data)
    else:
      node = self.head
      while node.next:
        node = node.next
      node.next = Node(data)

  def desc(self):
    node = self.head
    while node:
      print(node.data)
      node = node.next

  def delete(self, data):
    if self.head == "":
      print("해당 값을 가진 노드가 없습니다")
      return   
    if self.head.data == data:
      temp = self.head
      self.head = self.head.next
      del temp
    else:
      node = self.head
      while node.next:
        if node.next.data == data:
          temp = node.next
          node.next = node.next.next
          del temp
          return
          pass
        else:
          node = node.next

  def search_node(self, data):
    node = self.head
    while node:
      if node.data == data:
        return node
      else:
        node = node.next
        
node_mgmt = NodeMgmt(0)
for data in range(1, 10):
    node_mgmt.add(data)

node = node_mgmt.search_node(4)
print (node.data)